### PR TITLE
implement naive bayes map, mse, mae for CF

### PR DESCRIPTION
**Pull request by tiffanyj41**

### CODE REVIEW

File: src/algorithms/cf.py [hunk 1]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+    # create RDD for range of ratings, ie. [1, 2, 3, 4, 5] for ratings 1-5
+    min_rating = training_data.map(lambda (u,i,r): r).min()
+    max_rating = training_data.map(lambda (u,i,r): r).max()
+    range_of_ratings = sc.parallelize(list(range(int(min_rating), int(max_rating + 1))))
+
+    # since we have to determine the probability of rating r for each user and item, 
+    # we have to create a RDD with [(rating, (user, item))] for each rating
+    # ie. [(rating_1, (user, item)), (rating_2, (user, item)), (rating_3, (user, item)), ..., (rating_5, (user, item))]
+    ui = training_data.map(lambda (u,i,r): (u,i))
+    rCombo_ui = rangeOfRatings.cartesian(ui).map(lambda (r, (u,i)): (float(r), (u, i)))
```
Comment (bethke) [inline src/algorithms/cf.py]: should use range_of_ratings: rCombo_ui = rangeOfRatings.cartesian(ui).map(lambda (r, (u,i)): (float(r), (u, i)))

File: src/algorithms/cf.py [hunk 2]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+    # add probR to [(user_id, item_id, rating)]
+    components = rCombo_ui.join(probR)
+
+    # add probRU to [(user_id, item_id, rating, probR)]
+    tmp_a = components.map(lambda (r, ((u,i), prob_r)): ((u, r), (i, prob_r)))
+    tmp_b = probRU.map(lambda (u, r, prob_ru): ((u, r), prob_ru))
+    components = tmp_a.join(tmp_b)
+
+    # add probRI to [(user_id, item_id, rating, probR, probRU)]
+    tmp_a = components.map(lambda ((u, r), ((i, prob_r), prob_ru)): (i, r), (u, prob_r, prob_ru))
```
Comment (bethke) [inline src/algorithms/cf.py]: tmp_a = components.map(lambda ((u, r), ((i, prob_r), prob_ru)): ((i, r), (u, prob_r, prob_ru)))

You are missing a few parenthesis on the right side of the lambda

File: src/algorithms/cf.py (hunk 1 above)
Comment (tiffanyj41) [inline src/algorithms/cf.py]: good catch; making that fix now.

File: src/algorithms/cf.py (hunk 2 above)
Comment (tiffanyj41) [inline src/algorithms/cf.py]: Okay, will do.

File: src/algorithms/cf.py [hunk 3]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+    if computeFromScratch:
+        ui_allBayesProb = __calc_naive_bayes_components(training_data, sc)
+    else:
+        if ui_allBayesProb is None:
+            raise Exception("ERROR: ui_allBayesProb is not defined although user specified not to calculate ui_allBayesProb not from scratch.")
+
+    return ui_allBayesProb.mapValues(calculate_bayes_map)
+
+
+def calc_naive_bayes_mse(training_data, sc):
```
Comment (tiffanyj41) [inline src/algorithms/cf.py]: Noted that it requires additional arguments: (training_data, sc, computeFromScratch=True, ui_allBayesProb=None)

File: src/algorithms/cf.py [hunk 4]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+    if computeFromScratch:
+        ui_allBayesProb = __calc_naive_bayes_components(training_data, sc)
+    else:
+        if ui_allBayesProb is None:
+            raise Exception("ERROR: ui_allBayesProb is not defined although user specified not to calculate ui_allBayesProb not from scratch.")
+
+    return ui_allBayesProb.mapValues(calculate_bayes_mse)
+
+
+def calc_naive_bayes_mae(training_data, sc):
```
Comment (tiffanyj41) [inline src/algorithms/cf.py]: Noted that it requires additional arguments: (training_data, sc, computeFromScratch=True, ui_allBayesProb=None)

File: src/algorithms/cf.py [hunk 5]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+        predictions: predicted ratings of every user-item combination in the format of a RDD of [(userId, itemId, predictedRating)].
+    """
+
+    def calculate_bayes_mse(value):
+        predicted = 0.
+        for rating, bayes_prob in value:
+            predicted += rating * bayes_prob
+        return predicted
+
+    if computeFromScratch:
```
Comment (bethke) [inline src/algorithms/cf.py]: Need to pass in

File: src/algorithms/cf.py [hunk 6]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+        ui_allBayesProb = __calc_naive_bayes_components(training_data, sc)
+    else:
+        if ui_allBayesProb is None:
```
Comment (bethke) [inline src/algorithms/cf.py]: Need to pass in

File: src/algorithms/cf.py [hunk 7]
```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
...
+    we make the assumption that the latter part of the multiplication, ( (P(u) * P(i)) / P(u, i)), can be ignored.
+
+    Args:
+        training_data: the data used to train the RecSys algorithm in the format of a RDD of [ (userId, itemId, actualRating) ]
+
+    Returns:
+        predictions: predicted ratings of every user-item combination in the format of a RDD of [(userId, itemId, predictedRating)].
+    """
+
+    def calculate_bayes_mae(value):
```
Comment (bethke) [inline src/algorithms/cf.py]: Not sure if this should be a new function...there are some indentation issues it looks like

Comment (tiffanyj41) [inline src/algorithms/cf.py]: Yes, it is an indentation issue.

**Pull request merged**